File: leetcode/106/P106.rb
# 106. Construct Binary Tree from Inorder and Postorder Traversal
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} inorder
# @param {Integer[]} postorder
# @return {TreeNode}
def build_tree(inorder, postorder)
	$inorder = inorder
	$postorder = postorder
	build 0, inorder.size - 1, postorder.size - 1
end

def build i0, i1, p1
	return nil if i0 > i1
	node = TreeNode.new $postorder[p1]
	i = $inorder.index $postorder[p1]
	node.left = build i0, i - 1, p1 - i1 + i - 1
	node.right = build i + 1, i1, p1 - 1
	node
end

File: leetcode/106/src/P106.java
/**
 * Created by qwordy on 2/24/16.
 * 106. Construct Binary Tree from Inorder and Postorder Traversal
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public TreeNode buildTree(int[] inorder, int[] postorder) {
		this.inorder = inorder;
		this.postorder = postorder;
		return build(0, inorder.length - 1, postorder.length - 1);
	}

	private TreeNode build(int i0, int i1, int p1) {
		int i;
		if (i0 > i1) return null;
		TreeNode node = new TreeNode(postorder[p1]);
		for (i = i0; i <= i1; i++)
			if (inorder[i] == postorder[p1])
				break;
		node.left = build(i0, i - 1, p1 - i1 + i - 1);
		node.right = build(i + 1, i1, p1 - 1);
		return node;
	}

	int[] inorder, postorder;
}

public class P106 {
}
